core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -306,33 +334,43 @@ public LockResource acquireBlockReinitLockResource() {
    *
    * @param updateClusterConf whether cluster level configuration should be updated
    * @param updatePathConf whether path level configuration should be updated
-   * @throws TimeoutException when timed out during being blocked by ongoing RPCs
-   * @throws InterruptedException when the calling thread is interrupted
+   * @return whether reinitialization runs and succeeds
    * @throws UnavailableException when failed to load configuration from master
    * @throws IOException when failed to close the context
    */
-  public synchronized void reinit(boolean updateClusterConf, boolean updatePathConf)
-      throws TimeoutException, InterruptedException, UnavailableException, IOException {
-    try (LockResource r = mReinitializer.acquireWriteLockResource()) {
-      InetSocketAddress masterAddr;
-      try {
-        masterAddr = getMasterAddress();
-      } catch (IOException e) {
-        throw new UnavailableException("Failed to get master address during reinitialization", e);
+  public boolean reinit(boolean updateClusterConf, boolean updatePathConf)
+      throws UnavailableException, IOException {
+    if (!mReinitLock.writeLock().tryLock()) {
```

Comment:
With these changes the only chance for HashSync to happen is that there are no active RPCs at the time of heartbeat. That is pretty unfair to hash-sync. How about try-locking with a timeout value? You could use the same timeout that you used before. That way has-sync will have a better chance to run.

Reply:
but even if reinit acquires this lock, it still needs to acquire the counter resource to allow reinit. If there are streaming operations happening, reinit will not happen even if this write lock is acquired.

I changed to use a fair lock, and just lock instead of tryLock here, so that reinit has a better chance to run.